### ISSUE TITLE
Designate Required vs Optional Parameters/Options

### ISSUE THREAD

**Issue by bbrandt** [natemcmaster/CommandLineUtils#12]
Currently, it is up to the application developer using `CommandLineUtils` to validate command line input and display an appropriate error if an _Option_ that is required is left out.  I propose that `CommandLineUtils` take on this responsibility, allowing the library user to designate parameters as required and provide appropriate feedback to the user when the parameter is not provided.

This may be a nice feature to do after #8 so the user of the library need only add an attribute of [Required] to make this designation.

**Comment by natemcmaster** [comment]
I definitely like the idea of adding some sort customizable validation step. To implement #8, I will at least a little bit of this to validate I can convert a string to CLR types. I was planning to implement rudimentary validation first as a completely `internal` detail for now and then hash out an API in 2.2 or later.

But if #8 hypothetically never got implemented, do you have any thoughts on what a good API for this would look like?

**Comment by bbrandt** [comment]
Allow user to call a `RequiredOption()` for options that are required or `Option()` for options that are optional.  `RequiredOption()` would have the same set of overloads as `Option()`.

```
CommandOption optionalOption = Option("-o | --opt", "you can skip this if you like", CommandOptionType.SingleValue);
CommandOption requiredOption= RequiredOption("-r | --req", "you have to enter this", CommandOptionType.SingleValue);
```

The rest of the code for accessing the options would remain unchanged as they are all using the same CommandOption type.

During the `CommandLineApplication.Execute()`, CLU validates the arguments and does not execute the provided Func if a required option is missing.

I would like to work out a way to show all (maybe limited to 3-5) missing required arguments as well as the current functionality to fail on unrecognized arguments to the end user at the same time so they are not playing whack-a-mole with our command line apps, seeing one error at a time, adding a missing parameter, and re-running repeatedly with no end in sight.

Currently, the first unrecognized argument throws, only notifying the user of their first mistake.  What's the thought behind this behavior?

**Comment by georgiosd** [comment]
Why not just use the standard validation attributes (for the 2.1 attributes API)?

**Comment by natemcmaster** [comment]
Thought about it more, here is what I'm thinking. Would like your feedback.

## Attributes
Design:

Support any implementation of System.ComponentModel.DataAnnotations.ValidationAttribute. This includes `RequiredAttribute` and others.

Result:
When validation fails, the application exits with code 1 and prints an error message.

Usage:

```c#
class Program
{
    static int Main(string[] args) => CommandLineApplication.Execute<Program>(args);

    [Required]
    [Option(Description = "Required. The message")]
    private string Message { get; }

    [Required(ErrorMessage = "my custom error message")]
    [EmailAddress]
    [Option("--to <EMAIL>", Description = "Required. The recipient.")]
    public string To { get; }

    private void OnExecute(IConsole console)
    {
        console.WriteLine("To = " + To);
        console.WriteLine("Message = " + Message);
    }
}
```

## Builder api
Design:
Add extension methods that build of `CommandOption`, such as `.IsRequired()`. These should roughly map to the same values provided by `RequiredAttribute` and others.

Usage:
```c#
class Program
{
    static int Main(string[] args)
    {
        var app = new CommandLineApplication();

        var optionMessage = app.Option("-m|--message <MSG>", "Required. The message.", CommandOptionType.SingleValue)
            .IsRequired();

        var optionReceiver = app.Option("--to <EMAIL>", "Required. The recipient.", CommandOptionType.SingleValue)
            .IsRequired()
            .IsEmailAddress();

        app.OnExecute(() =>
        {
            Console.WriteLine("To = " + optionReceiver.Value());
            Console.WriteLine("Message = " + optionMessage.Value());
        });

        return app.Execute(args);
    }
}
```

**Comment by georgiosd** [comment]
That looks great! I'd go a step further and print the validation requirements for each option at the end.

Like: The recipient (Required, Email)

**Comment by natemcmaster** [comment]
I merged #20 which includes support for designating required arguments/options.

The validation implementation needs more refinement. But since the original ask for this issue was only for required/optional parameters, I'm going to close this as "done" for now. You can test our a pre-release version of this API with this build: https://www.myget.org/feed/natemcmaster/package/nuget/McMaster.Extensions.CommandLineUtils/2.1.0-rc.99+master

Let's use https://github.com/natemcmaster/CommandLineUtils/issues/24 to track the refinements and improvements to parameter validation.

**Comment by georgiosd** [comment]
Thanks man, that looks great!

**Comment by bbrandt** [comment]
Great API!  Good idea using System.ComponentModel.DataAnnotations.  Love that we get EmailAddress and other ValidationAttribute implementations basically for free.

**Comment by natemcmaster** [comment]
This is now available in 2.1.0-rc. 

https://www.nuget.org/packages/McMaster.Extensions.CommandLineUtils/2.1.0-rc
